hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/PrngSystemPrecompiledContract.java

```diff
@@ -0,0 +1,285 @@
+	private final GlobalDynamicProperties properties;
+	private long gasRequirement;
+
+	@Inject
+	public PrngSystemPrecompiledContract(
+			final GasCalculator gasCalculator,
+			final PrngLogic prngLogic,
+			final ExpiringCreations creator,
+			final SideEffectsTracker tracker,
+			final RecordsHistorian recordsHistorian,
+			final PrecompilePricingUtils pricingUtils,
+			final LivePricesSource livePricesSource,
+			final GlobalDynamicProperties properties) {
+		super(PRECOMPILE_NAME, gasCalculator);
+		this.prngLogic = prngLogic;
+		this.creator = creator;
+		this.tracker = tracker;
+		this.recordsHistorian = recordsHistorian;
+		this.pricingUtils = pricingUtils;
+		this.livePricesSource = livePricesSource;
+		this.properties = properties;
+	}
+
+	@Override
+	public long gasRequirement(final Bytes bytes) {
+		return gasRequirement;
+	}
+
+	@Override
+	public PrecompileContractResult computePrecompile(final Bytes input, final MessageFrame frame) {
+		final var gasNeeded = calculateGas(Instant.ofEpochSecond(frame.getBlockValues().getTimestamp()));
+		final var result = computePrngResult(gasNeeded, input, frame);
+
+		if (frame.isStatic()) {
+			final var proxyUpdater = (HederaStackedWorldStateUpdater) frame.getWorldUpdater();
+			if (!proxyUpdater.isInTransaction()) {
+				// This thread is answering a ContractCallLocal query; don't create a record or change
+				// instance fields, just return the gas required and output for the given input
+				return result.getLeft();
+			}
+		}
+		final var randomNum = result.getLeft().getOutput();
+		final var childRecord = result.getRight() == null ?
+				createSuccessfulChildRecord(randomNum, frame, input) :
+				createUnsuccessfulChildRecord(result.getRight(), frame);
+
+		final var parentUpdater = updater.parentUpdater();
+		if (parentUpdater.isPresent()) {
+			final var parent = (AbstractLedgerWorldUpdater) parentUpdater.get();
+			parent.manageInProgressRecord(recordsHistorian, childRecord, body(randomNum, input));
+		} else {
+			throw new InvalidTransactionException("PRNG precompile frame had no parent updater", FAIL_INVALID);
+		}
+
+		return result.getLeft();
+	}
+
+	Pair<PrecompileContractResult, ResponseCodeEnum> computePrngResult(
+			final long gasNeeded,
+			final Bytes input,
+			final MessageFrame frame) {
+		try {
+			validateTrue(frame.getRemainingGas() >= gasNeeded, INSUFFICIENT_GAS);
+			final var randomNum = generatePseudoRandomData(input);
+			return Pair.of(PrecompiledContract.PrecompileContractResult.success(randomNum), null);
+		} catch (InvalidTransactionException e) {
+			return Pair.of(PrecompiledContract.PrecompileContractResult.halt(null,
+					Optional.ofNullable(ExceptionalHaltReason.INVALID_OPERATION)), e.getResponseCode());
+		} catch (IllegalArgumentException e) {
+			return Pair.of(PrecompiledContract.PrecompileContractResult.halt(null,
+					Optional.ofNullable(ExceptionalHaltReason.INVALID_OPERATION)), INVALID_PRNG_RANGE);
```

Comment:
Q: Is range the only issue that could cause this exception?
If not it could lead to misleading troubleshooting details

Reply:
yes, when the range is outside Integer range, `WORD_DECODER.decode` causes the error.

Reply:
Okay